cloudevents/pydantic/json_format.py

```diff
@@ -0,0 +1,64 @@
+import base64
+import json
+import re
+import typing
+
+_NON_BINARY_JSON_DATA_KEY = "data"
+_BINARY_JSON_DATA_KEY = "data_base64"
+
+# If the type of data is not Binary, the implementation will next determine whether
+# the value of the datacontenttype attribute declares the data to contain JSON-formatted content.
+# Such a content type is defined as one having a media subtype equal to json or ending with a +json format extension.
+# That is, a datacontenttype declares JSON-formatted content if its media type, when stripped of parameters,
+# has the form */json or */*+json. If the datacontenttype is unspecified, processing SHOULD proceed as
+# if the datacontenttype had been specified explicitly as application/json.
+_JSON_CONTENT_TYPE_PATTERN = re.compile(
+    r"^.+/(.+\+)?json.*$", re.UNICODE | re.IGNORECASE
+)
+
+
+def _is_json_content_type(obj: typing.Dict[str, typing.Any]) -> bool:
+    data_content_type = obj.get("datacontenttype")
+    if data_content_type is None:
+        data_content_type = "application/json"  # allowed to assume json by spec
+    return bool(_JSON_CONTENT_TYPE_PATTERN.match(data_content_type))
+
+
+class NonJsonContentTypeData(ValueError):
+    def __init__(self):
+        super(NonJsonContentTypeData, self).__init__(
+            "Given data is not a json content type, event though the datacontenttype attribute claims otherwise"
+        )
+
+
+def ce_json_dumps(
+    obj: typing.Dict[str, typing.Any], *args, **kwargs
+) -> typing.Union[str, bytes]:
+    data = obj.get(_NON_BINARY_JSON_DATA_KEY)
+    if data is not None:
+        if isinstance(data, (str, bytes)) and _is_json_content_type(obj):
+            try:
+                data = ce_json_loads(data)
+            except UnicodeDecodeError:
+                raise NonJsonContentTypeData()
+            obj[_NON_BINARY_JSON_DATA_KEY] = data
+        if isinstance(data, (bytes, bytes, memoryview)):
+            obj[_BINARY_JSON_DATA_KEY] = base64.b64encode(data).decode("ascii")
+            del obj[_NON_BINARY_JSON_DATA_KEY]
+    return json.dumps(obj, *args, **kwargs)
+
+
+def ce_json_loads(*args, **kwargs):
+    result = json.loads(*args, **kwargs)
+    if isinstance(result, dict):
+        if _BINARY_JSON_DATA_KEY in result:
+            result[_NON_BINARY_JSON_DATA_KEY] = base64.b64decode(
+                result.pop(_BINARY_JSON_DATA_KEY)
+            )
+        data = result.get(_NON_BINARY_JSON_DATA_KEY)
+        if isinstance(data, (str, bytes)) and _is_json_content_type(result):
+            try:
+                result[_NON_BINARY_JSON_DATA_KEY] = ce_json_loads(data)
+            except UnicodeDecodeError:
+                raise NonJsonContentTypeData()
+    return result
```

Comment:
These 2 methods are pretty hard to grasp. Also, the docs are missing. Ideally, you should be able to just use an existing encoding. Or, maybe decode the Pydantic model into a standard CloudEvent and then use the existing API.

Reply:
I agree, these are the first functions im converting. 

Did not thought about converting into into an http cloudevent
i'm override the json_loads in the pydantic config. these are internal use only functions. but seems that your proposal is much simpler